src/irmin-pack/unix/gc.ml

```diff
@@ -83,39 +73,89 @@ module Worker = struct
       in
       aux off len
 
-    (** [iter_from_node_key node_key _ _ ~f] calls [f] with the key of the node
-        and iterates over its children.
+    module Priority_queue = struct
+      module Offset_rev = struct
+        type t = int63
+
+        let equal = Int63.equal
+        let hash (t : t) = Hashtbl.hash t
+        let compare a b = Int63.compare b a
+      end
+
+      module Table = Hashtbl.Make (Offset_rev)
+      module Pq = Binary_heap.Make (Offset_rev)
+
+      type 'a t = { pq : Pq.t; marks : 'a Table.t }
+
+      let create size =
+        { pq = Pq.create ~dummy:Int63.zero size; marks = Table.create size }
+
+      let is_empty t = Pq.is_empty t.pq
 
-        [f k] returns [Follow] or [No_follow], indicating the iteration
-        algorithm if the children of [k] should be traversed or skiped. *)
-    let iter node_key node_store ~f k =
-      let marks = Table.create 1024 in
-      let mark offset = Table.add marks offset () in
-      let has_mark offset = Table.mem marks offset in
-      let rec iter_from_node_key_exn node_key node_store ~f k =
+      let pop { pq; marks } =
+        let elt = Pq.pop_minimum pq in
+        let payload = Table.find marks elt in
+        Table.remove marks elt;
+        (elt, payload)
+
+      let add { pq; marks } elt payload =
+        if not (Table.mem marks elt) then (
+          Table.add marks elt payload;
+          Pq.add pq elt)
+    end
+
+    (** [iter_reachable commit _ ~f] calls [f ~off ~len] once for each [offset]
+        and [length] of the reachable tree objects and immediate parent commits
+        from [commit]. *)
+    let iter_reachable commit node_store ~f =
+      let todos = Priority_queue.create 1024 in
+      let rec loop () =
+        if not (Priority_queue.is_empty todos) then (
+          let offset, (length, key_opt) = Priority_queue.pop todos in
+          f ~off:offset ~len:length;
+          (match key_opt with None -> () | Some key -> iter_node key);
```

Comment:
what do you think of reusing a follow, no follow variant here instead of the option type?